File: mcp-code-intelligence-nodejs/src/analyzers/complexity/counters/PythonCounter.ts
/**
 * KSA-161: Python decision point counter.
 */

import type { SyntaxNode } from '../../../parsers/types.js';
import { BaseNodeCounter } from './BaseNodeCounter.js';

export class PythonCounter extends BaseNodeCounter {
  readonly language = 'python';
  readonly branchNodeTypes = [
    'if_statement',
    'elif_clause',
    'conditional_expression',  // ternary: x if cond else y
  ];
  readonly loopNodeTypes = [
    'for_statement',
    'while_statement',
  ];
  readonly logicalOperators = ['and', 'or'];
  readonly exceptionNodeTypes = ['except_clause'];

  protected isLogicalOp(node: SyntaxNode): boolean {
    if (node.type === 'boolean_operator') {
      const op = node.childForFieldName('operator');
      if (op && this.logicalOperators.includes(op.text)) return true;
      // Fallback: check children for 'and'/'or' keywords
      for (let i = 0; i < node.childCount; i++) {
        const child = node.child(i);
        if (child && (child.text === 'and' || child.text === 'or')) return true;
      }
    }
    return false;
  }

  protected getReturnNodeTypes(): string[] {
    return ['return_statement'];
  }
}

File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/complexity/counters/PythonCounter.kt
/**
 * KSA-161: Python AST node counter.
 */
package com.codeintel.analyzers.complexity.counters

import com.codeintel.analyzers.complexity.BaseNodeCounter

class PythonCounter : BaseNodeCounter() {
    override val language = "python"
    override val branchNodeTypes = setOf(
        "if_statement", "elif_clause", "conditional_expression",
    )
    override val loopNodeTypes = setOf(
        "for_statement", "while_statement",
    )
    override val logicalOperators = setOf("and", "or")
    override val exceptionNodeTypes = setOf("except_clause")

    override fun getReturnNodeTypes(): Set<String> =
        setOf("return_statement")
}
